### PR TITLE
ENH: adds  bind kwarg to odo

### PR DESCRIPTION
**Pull request by llllllllll**

### CODE REVIEW

**Comment by cpcloud** [comment]
Can you add a test for converting something without bound metadata, e.g., a function (to reproduce the case that motivated this PR).

**Comment by cpcloud** [comment]
Can you also add some documentation to the effect of "Hey be aware of the bound/unbound metadata issue, and here the two different ways that this works and in certain cases you *must* pass `bind` in"

**Comment by llllllllll** [comment]
I changed the tests to be paramaterized on bind so that they will run once with bound objects and then once with unbound objects and bind passed to odo.

File: docs/source/sql.rst [hunk 1]
```diff
@@ -50,6 +53,64 @@ SQL uris consist of the following
...
+.. code-block:: python
+
+   >>> import sqlalchemy as sa
+   >>> metadata = sa.MetaData(bind='sqlite:///db.db')  # NOTE: pass the uri to the db here
+   >>> tbl = sa.Table(
+   ...     'tbl',
+   ...     metadata,
+   ...     sa.Column('a', sa.Integer, primary_key=True),
+   ... )
+   >>> odo([[1], [2], [3]], tbl)  # this know knows where to fild the table.
```
Comment (cpcloud) [inline docs/source/sql.rst]: typo in `fild`

File: odo/backends/sql.py [hunk 2]
```diff
@@ -455,29 +465,28 @@ def append_anything_to_sql_Table(t, o, **kwargs):
...
 @append.register(sa.Table, sa.sql.Select)
-def append_select_statement_to_sql_Table(t, o, **kwargs):
-    if not o.bind == t.bind:
-        return append(t, convert(Iterator, o, **kwargs), **kwargs)
-
-    assert o.bind.has_table(t.name, t.schema), \
+def append_select_statement_to_sql_Table(t, o, bind=None, **kwargs):
+    tbind = getbind(t, bind)
+    obind = getbind(o, bind)
+    if not tbind == obind:
```
Comment (cpcloud) [inline odo/backends/sql.py]: can you change this to `tbind != obind`

File: odo/backends/sql.py [hunk 3]
```diff
@@ -455,29 +465,28 @@ def append_anything_to_sql_Table(t, o, **kwargs):
...
+            **kwargs
+        )
+
+    assert obind.has_table(t.name, t.schema), \
         'tables must come from the same database'
 
     query = t.insert().from_select(o.columns.keys(), o)
 
-    with o.bind.connect() as conn:
+    with obind.connect() as conn:
```
Comment (cpcloud) [inline odo/backends/sql.py]: maybe this should be `bind` just for clarity? i realize that if `bind is not None` then `obind == bind`, but IMO it reads better if you use `bind` here.

File: docs/source/sql.rst [hunk 4]
```diff
@@ -50,6 +53,64 @@ SQL uris consist of the following
 * A table name with the ``::`` separator:  ``::accounts``
 
 
+Binds
```
Comment (cpcloud) [inline docs/source/sql.rst]: Maybe call this `Bound vs. Unbound MetaData`

Comment (cpcloud) [inline docs/source/sql.rst]: Or, `Executing Odo Against Databases`

**Comment by llllllllll** [comment]
Does this look good to merge?

**Comment by cpcloud** [comment]
Let me try this branch out on some internal stuff before merging. Should be okay to merge first thing tomorrow or later tonight if I can get to it

**Comment by llllllllll** [comment]
Sounds good.

**Comment by cpcloud** [comment]
@llllllllll Thanks! This is in!

**Comment by llllllllll** [comment]
hyphy

**Pull request merged**